kitsune/wiki/models.py

```diff
@@ -263,39 +279,53 @@ def save(self, *args, **kwargs):
         # which would cause a rollback, which would negate any category changes
         # we make here, so don't worry:
         self._clean_category()
+        self._clean_template_status()
 
         slug_changed = hasattr(self, 'old_slug')
         title_changed = hasattr(self, 'old_title')
 
-        # If the slug changed, we clear out the share link so it gets regenerated.
-        self.share_link = ''
+        templateness_changed = (
+            title_changed and
+            # If the title was a template before, and is not now, or vice-versa
+            (self.title.startswith(TEMPLATE_TITLE_PREFIX) !=
+             self.old_title.startswith(TEMPLATE_TITLE_PREFIX)))
+
+        if slug_changed:
+            # Clear out the share link so it gets regenerated.
+            self.share_link = ''
 
         super(Document, self).save(*args, **kwargs)
 
         # Make redirects if there's an approved revision and title or slug
         # changed. Allowing redirects for unapproved docs would (1) be of
         # limited use and (2) require making Revision.creator nullable.
-        if self.current_revision and (slug_changed or title_changed):
+        #
+        # If the "templateness" changed, don't try and make a redirect.
+        # The category and the title-prefix have to stay consistent, and
+        # redirects are in the same category as their targets. There is not
+        # a clear way to construct the redirect without invalidating one of
+        # those conditions.
+        if self.current_revision and (slug_changed or title_changed) and not templateness_changed:
```

Comment:
I'm not following this at all. It seems like the additional `not templateness_changed` only affects the case where the document got switched to a template or the template got switched to a document. Is that what you want? Do we want redirects for templates at all?